<a href="https://colab.research.google.com/github/Bhuvan23-A/flyrank_ml_intern/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

#Structured Content Archetype Clustering

## Why:
It would be really hard to find a strategy which would work on all the webpages, but there it is probable that a strategy would work on similar clusters of data. As we do not have predefined classes we have to use kmeans algo for this.

## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

##Decision:
It makes their life easier to work on similar bits rather than the whole
##Who acts on it and the action?:
SEO manager and content editors to devise strategies or plans to improve pages which have low views and CTR and also to get an idea what makes a page get high views or low views i.e to classify them.
So that they can make more informed decision on new pages.
## What does a wrong call cost?:-
If a page is under a cluster in which it does not belong, we can get a wrong ideas on what works and doesn't. Also we the overall plan may get disturbed.
## Why ML:-
We have large amount of pages and we do not have predefined classes. Kmeans would solve both these problems and it also works on multidimensional interaction which is a plus

## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [6]:
import pandas as pd
df = pd.read_csv("/content/content_refresh_anonymized.csv")

df.head()

total_rows = len(df)
total_clients = df["client_id"].nunique()
# 1. Archetype 1: Hidden Gems (Page 1 rank, high impressions, low CTR)
hidden_gems = df[
    (df["avg_position"] > 0)
    & (df["avg_position"] <= 10)
    & (df["impressions_90d"] >= 500)
    & (df["ctr"] < 1.0)
]
# 2. Archetype 2: High-Demand Decaying Assets
declining_demand = df[
    (df["impressions_90d"] >= 500) & (df["trend_direction"] == "down")
]
impressions_at_risk = declining_demand["impressions_90d"].sum()
# 3. Archetype 3: Dead-Weight / Low Demand Tail
low_demand = df[df["impressions_90d"] < 10]
print("=== STARTER DATA EVIDENCE ===")
print(
    f"Total Content Items (Grain = 1 page): {total_rows:,} across"
    f" {total_clients} clients\n"
)
print(
    f"1. Hidden Gems (Page 1 rank, >=500 imp, CTR < 1.0%): {len(hidden_gems):,}"
    f" pages ({len(hidden_gems)/total_rows*100:.1f}%)"
)
print(
    "2. High-Demand Assets in Active Decline (>=500 imp, trend 'down'):"
    f" {len(declining_demand):,} pages"
)
print(
    f"   -> Total 90-day impressions at risk in this group:"
    f" {impressions_at_risk:,}"
)
print(
    f"3. Low/Zero Demand Tail (<10 imp in 90 days): {len(low_demand):,} pages"
    f" ({len(low_demand)/total_rows*100:.1f}%)"
)

=== STARTER DATA EVIDENCE ===
Total Content Items (Grain = 1 page): 30,000 across 32 clients

1. Hidden Gems (Page 1 rank, >=500 imp, CTR < 1.0%): 7,188 pages (24.0%)
2. High-Demand Assets in Active Decline (>=500 imp, trend 'down'): 9,961 pages
   -> Total 90-day impressions at risk in this group: 79,042,325
3. Low/Zero Demand Tail (<10 imp in 90 days): 3,746 pages (12.5%)


Inspecting the 30,000 pages in the starter dataset reveals strong evidence of distinct behavioral archetypes:

Inventory Scale: There are 30,000 pages across 32 clients, making manual review impossible.

Hidden Gems (24.0%): 7,188 pages rank in Google's Top 10 with
≥
500
≥500 impressions, but have a CTR below
1.0
%
1.0%. These represent low-hanging fruit where snippet optimizations can unlock clicks without changing the article body.

Decaying Winners: 9,961 pages have high demand (
≥
500
≥500 impressions) but are actively trending downward, putting over 79 million impressions at risk if not refreshed.

Low-Demand Tail (12.5%): 3,746 pages receive fewer than 10 impressions over 90 days and represent potential pruning or consolidation candidates.

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

## What our work CAN claim:
->Observed Multivariate Clusters: We can claim that mathematical clustering (e.g., K-Means) groups content items into stable, observable archetypes based on multivariate search metrics (impressions, position, CTR, engagement, and age).

->Directional Decision Support: We can claim that these clusters provide an operational triage framework to help editorial teams prioritize their limited time.

-> Empirical Association: We can describe how features correlate with cluster assignment within this historical snapshot dataset.
## What our work CANNOT claim:
->No "Semantic" Clustering: We cannot claim our clusters represent semantic meaning or topic understanding, because our dataset contains structured metrics and metadata, not raw article text.

->No Causal Guarantees: We cannot claim that updating an article will cause it to recover its ranking or traffic. Proving causality requires controlled A/B experiments.

->No Reverse-Engineering Google: We cannot claim to have discovered Google's internal ranking algorithm factors.

->No Truth in Product Rules: We do not treat proprietary product rules like health_score as ground truth; our analysis is grounded solely in observable signals.

## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.